### ISSUE TITLE
0.9.16 crash octoprint startup 

### ISSUE THREAD

**Issue by devildant** [jneilliii/OctoPrint-TPLinkSmartplug#102]
Hello, 
after the update of tp link, impossible to restart octoprint, there is a stacktrace at startup
```
2019-05-01 23:25:18,803 - octoprint.cli.server - ERROR - Uncaught exception
Traceback (most recent call last):
  File "/home/pi/oprint/bin/octoprint", line 11, in <module>
    load_entry_point('OctoPrint==1.3.11rc3', 'console_scripts', 'octoprint')()
  File "/home/pi/oprint/local/lib/python2.7/site-packages/octoprint/__init__.py", line 609, in main
    octo(args=args, prog_name="octoprint", auto_envvar_prefix="OCTOPRINT")
  File "/home/pi/oprint/local/lib/python2.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/pi/oprint/local/lib/python2.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/pi/oprint/local/lib/python2.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/pi/oprint/local/lib/python2.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/pi/oprint/local/lib/python2.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/pi/oprint/local/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/pi/oprint/local/lib/python2.7/site-packages/octoprint/cli/server.py", line 206, in serve_command
    ignore_blacklist)
  File "/home/pi/oprint/local/lib/python2.7/site-packages/octoprint/cli/server.py", line 118, in run_server
    octoprint_server.run()
  File "/home/pi/oprint/local/lib/python2.7/site-packages/octoprint/server/__init__.py", line 478, in run
    extra=dict(plugin=plugin._identifier))
UnboundLocalError: local variable 'plugin' referenced before assignment


```
I try several times, without success, I have downgrader in 0.9.13

**Comment by devildant** [comment]
octoprint version 1.3.11rc3

**Comment by jneilliii** [comment]
Did you try deleting plugin's settings in config.yaml? None of the error log is referencing my plugin files, so it's hard to debug this one.

**Comment by devildant** [comment]
I tried, but nothing to do impossible to start octoprint, I am forced to go into safe mode.
remove tplink 0.9.16, and then restart normally and reinstall 0.9.13.
with the 0.9.13 no problem

**Comment by devildant** [comment]
I doubt that it will be easy to fix, I have no more info unfortunately.

**Comment by jneilliii** [comment]
Ok, I'm able to reproduce on my end with the upgrade from 0.9.13 to 0.9.16.  I have reverted that release for now and will investigate further.

**Comment by devildant** [comment]
I wish you good luck

**Comment by jneilliii** [comment]
@devildant, can you try installing the upgrade from the url below and let me know if it gets through for you or not.  It seems to be working for me in my testing.

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16a.zip

**Comment by devildant** [comment]
hello, 
it's ok good jobs.

I have now detected a JS problem with the latest version of tplink : 
![image](https://user-images.githubusercontent.com/8700106/57062865-ce787700-6cc1-11e9-9154-db6d0f1d104b.png)

problem with lib :
 // source: plugin/errortracking/js/sentry.min.js
/*! @sentry/browser 5.0.5 (913eafc4) | https://github.com/getsentry/sentry-javascript */

**Comment by jneilliii** [comment]
During upgrade ther plugin was trying to create the sqlite3 database using a getplugindatafolder call,  which isn't available I guess during that phase of loading, so I moved that code to a different section. 

Sentry is part of the new error tracking, not my plugin.

**Comment by devildant** [comment]
> During upgrade ther plugin was trying to create the sqlite3 database using a getplugindatafolder call, which isn't available I guess during that phase of loading, so I moved that code to a different section.

ok i see, thx

> Sentry is part of the new error tracking, not my plugin.
yes i know that sentry is not inside tplink plugin but with the 0.9.13 i don't have any error inside the console.
this error only appears if 0.9.16 is installed

**Comment by jneilliii** [comment]
Ok, I'll check it out to see if it's the same on my end. May be something that will require @foosel assistance.

**Comment by jneilliii** [comment]
Ok, I see what's going on.  Not sure what the sentry reference you posted is, but the errors are on the initial load of the page as the select list of the tab is getting populated. It triggers a change event of the list, which in turn is trying to graph the data.  I just need to add some error checking within that part of the graphing and it should resolve the issue.

**Comment by jneilliii** [comment]
Took a slightly different approach on the graphing issue.  In lieu of using change events, added an update button to the tab. This should resolve the js errors.  Give it a try from this link and let me know how it goes. Makes more sense that way as it won't make as many calls back to the database/api side. I'm sure there could still be some improvement to the tab interface in some way.
  
https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16.zip

**Comment by devildant** [comment]
good job, it works, for the automatic update of the graph, it might be interesting to rely on polling if it is enabled, right?

I also take the opportunity to point out another error that seems to have occurred at random with the previous version, and that I still have today

```
packed_plugins.js?70109bc7:9424 Uncaught TypeError: Cannot read property 'plugins' of undefined
    at tplinksmartplugViewModel.self.onEventSettingsUpdated (packed_plugins.js?70109bc7:9424)
    at packed_core.js?79f3153d:13843
    at Pn (packed_libs.js?ea9fbc9a:11208)
    at Function.<anonymous> (packed_libs.js?ea9fbc9a:11229)
    at packed_core.js?79f3153d:13841
    at DataUpdater.self._ifInitialized (packed_core.js?79f3153d:13886)
    at DataUpdater.self._onEvent (packed_core.js?79f3153d:13746)
    at packed_client.js?68a8d0b0:390
    at Pn (packed_libs.js?ea9fbc9a:11208)
    at Function.<anonymous> (packed_libs.js?ea9fbc9a:11229)
    at OctoPrintSocketClient.propagateMessage (packed_client.js?68a8d0b0:389)
    at packed_client.js?68a8d0b0:484
    at packed_libs.js?ea9fbc9a:11225
    at _t (packed_libs.js?ea9fbc9a:11213)
    at packed_libs.js?ea9fbc9a:11224
    at Function.<anonymous> (packed_libs.js?ea9fbc9a:11229)
    at SockJS.onMessage [as onmessage] (packed_client.js?68a8d0b0:483)
    at SockJS.EventTarget.dispatchEvent (packed_libs.js?ea9fbc9a:22585)
    at packed_libs.js?ea9fbc9a:23298
    at Array.forEach (<anonymous>)
    at SockJS._transportMessage (packed_libs.js?ea9fbc9a:23296)
    at WebSocketTransport.EventEmitter.emit (packed_libs.js?ea9fbc9a:22501)
    at WebSocketTransport.ws.onmessage (packed_libs.js?ea9fbc9a:25372)
```
the error is here : 

```
self.onEventSettingsUpdated = function(payload) {
        			   self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
        		        }
```

**Comment by jneilliii** [comment]
If I remember correctly, yes if polling is enabled the graph will update,  also on every percent complete change while printing. 

As for this other error, that's standard function calls documented all over the development examples. I'm not doing anything out of the ordinary there. When the  error happens does it effect functionality in any way?

**Comment by devildant** [comment]
> If I remember correctly, yes if polling is enabled the graph will update, also on every percent complete change while printing.
the graph only updates if you click on update, even if polling is enabled

> As for this other error, that's standard function calls documented all over the development examples. I'm not doing anything out of the ordinary there. When the error happens does it effect functionality in any way?

I did not detect any anomaly,  may be just put an if (it's my cracked side of dev :p)
```
self.onEventSettingsUpdated = function(payload) {
          if (self.settings.settings.hasOwnProperty("plugins") && self.settings.settings.plugins.hasOwnProperty("tplinksmartplug") ) 
             self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
}
```

I also found a bug with the graph menu and the octoprint tab (see photo)
![image](https://user-images.githubusercontent.com/8700106/57155071-c3653a00-6dda-11e9-80e1-3873670b4963.png)

for fix add this just after your div : <div class="row-fluid" id="tplinksmartplug_energy_graph" data-bind="visible: loginState.isUser()"></div>
```
<style>
.js-plotly-plot .plotly .modebar {
 z-index: 100!important;
}
</style>
```

last point promised,
I discovered a little problem with polling.
I wanted to put 0.5 in value, and it is not a good idea because it makes a fork bomb :p
for fix : 
```
if (parseInt(self.settings.settings.plugins.tplinksmartplug.pollingEnabled(),10)) {
				setTimeout(function() {self.checkStatuses();}, (parseInt(self.settings.settings.plugins.tplinksmartplug.pollingInterval(),10) * 60000));
			};

```

after I would like to be able to poll every 10 second :p

I use a lot of your plugin, which is why I bother you a bit.
and then my plugin (shutdown printer) uses tplink :)
in any case thanks for the work

**Comment by jneilliii** [comment]
Thanks, I'll look into this tonight.

**Comment by jneilliii** [comment]
Thanks for the heads up on the modebar style.  I added that to the css file.  I wouldn't have caught that one because I use my TabOrder plugin and change all the tabs to icons.

![image](https://user-images.githubusercontent.com/5249455/57169107-dffd7400-6dd2-11e9-87e9-abde4034fc11.png)

Anyway, I just updated the pre-release 0.9.16 version. The polling and print percentage increase should update the graph now.  It was recording the data but not telling the UI to update.  Install using the below url.

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16.zip

**Comment by devildant** [comment]
i will test that :) 
you could look for the fork bomb?

**Comment by devildant** [comment]
almost, it's plugins which is undefined
try this : 
```
self.onEventSettingsUpdated = function(payload) {
        			if (self.settings.settings.hasOwnProperty("plugins")) {
        				self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
        			}
        		}
```

**Comment by jneilliii** [comment]
> you could look for the fork bomb?
Not sure what this is, but I added in your proposed fix.

**Comment by jneilliii** [comment]
Just updated the pre-release version again with the onEventSettingsUpdated fix.

**Comment by devildant** [comment]
Hi,
The problem of error is more complex than I thought.
it's a problem of asynchronicity.
the onEventSettingsUpdated method must be called just after the method that ends "self.setting ...".
this last method must be asynchronous.
I am looking for a solution, I will send you the final fix

**Comment by devildant** [comment]
here below, the implementation that works (tested)
```
		self.onEventSettingsUpdated = function(payload) {
			if (self.settings.settings !== undefined && self.settings.settings.plugins !== undefined) {
				self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
			} 
		}
```

**Comment by jneilliii** [comment]
Thanks. I'll update later today as it seems the update graph on percent complete increases aren't working either and I need to figure out why.

**Comment by devildant** [comment]
on my side the chart, seems to be updated, each polling

**Comment by jneilliii** [comment]
Yeah it's working for polling, but not for print progress.

**Comment by devildant** [comment]
ha ok I did not try

**Comment by jneilliii** [comment]
Ok, I have the code updated and everything seems to be working now.  I ended up changing the knockout binding around a little, so please give some thorough testing if you get a chance and let me know how it goes.  

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16.zip

**Comment by devildant** [comment]
ok, i detected several problem sorry.
1) 
```
packed_core.js?79f3153d:4194 Uncaught TypeError: Cannot read property 'is' of undefined
    at SettingsViewModel.self.isDialogActive (packed_core.js?79f3153d:4194)
    at SettingsViewModel.self.onEventSettingsUpdated (packed_core.js?79f3153d:4733)
    at packed_core.js?79f3153d:13843
    at Pn (packed_libs.js?ea9fbc9a:11208)
    at Function.<anonymous> (packed_libs.js?ea9fbc9a:11229)
    at packed_core.js?79f3153d:13841
    at DataUpdater.self._ifInitialized (packed_core.js?79f3153d:13886)
    at DataUpdater.self._onEvent (packed_core.js?79f3153d:13746)
    at packed_client.js?68a8d0b0:390
    at Pn (packed_libs.js?ea9fbc9a:11208)
    at Function.<anonymous> (packed_libs.js?ea9fbc9a:11229)
    at OctoPrintSocketClient.propagateMessage (packed_client.js?68a8d0b0:389)
    at packed_client.js?68a8d0b0:484
    at packed_libs.js?ea9fbc9a:11225
    at _t (packed_libs.js?ea9fbc9a:11213)
    at packed_libs.js?ea9fbc9a:11224
    at Function.<anonymous> (packed_libs.js?ea9fbc9a:11229)
    at SockJS.onMessage [as onmessage] (packed_client.js?68a8d0b0:483)
    at SockJS.EventTarget.dispatchEvent (packed_libs.js?ea9fbc9a:22585)
    at packed_libs.js?ea9fbc9a:23298
    at Array.forEach (<anonymous>)
    at SockJS._transportMessage (packed_libs.js?ea9fbc9a:23296)
    at WebSocketTransport.EventEmitter.emit (packed_libs.js?ea9fbc9a:22501)
    at WebSocketTransport.ws.onmessage (packed_libs.js?ea9fbc9a:25372)
```
```
        self.isDialogActive = function() {
            return self.settingsDialog.is(":visible");
        };

```

2) Current power stay to 0
![image](https://user-images.githubusercontent.com/8700106/57184098-95edbe80-6eb6-11e9-9bde-ad339f10f334.png)


3) if we reload the page (by removing the cache) the status of the plugin is off even if it is on.

**Comment by devildant** [comment]
little gift for you, for compatibility with touchui plugin

for this render 
![image](https://user-images.githubusercontent.com/8700106/57186090-508eb880-6ed9-11e9-8c71-5fc7eeb977c5.png)

use this css
```
#touch body #tabs #tab_plugin_tplinksmartplug_link a:before, #touch body #navbar #all_touchui_settings #tab_plugin_tplinksmartplug_link a:before {
    content: "\f080";
}


#touch body #tabs #sidebar_plugin_tplinksmartplug_wrapper_link a:before, #touch body #navbar #all_touchui_settings #sidebar_plugin_tplinksmartplug_wrapper_link a:before {
    content: "\f1e6";
}


#touch body #sidebar_plugin_tplinksmartplug > div.accordion-inner > #tplinksmartplugs > div > div.on {
    color: #FFFFFF;
}

```

for this render : 
![image](https://user-images.githubusercontent.com/8700106/57186095-89c72880-6ed9-11e9-98ed-548ea7903363.png)

```
#touch body #tabs #tab_plugin_tplinksmartplug_link a:before, #touch body #navbar #all_touchui_settings #tab_plugin_tplinksmartplug_link a:before {
    content: "\f1e6\f080";
}

#touch body #tabs #sidebar_plugin_tplinksmartplug_wrapper_link a:before, #touch body #navbar #all_touchui_settings #sidebar_plugin_tplinksmartplug_wrapper_link a:before {
    content: "\f1e6";
}

#touch body #sidebar_plugin_tplinksmartplug > div.accordion-inner > #tplinksmartplugs > div > div.on {
    color: #FFFFFF;
}
```

**Comment by devildant** [comment]
and for button plug
add the span with name plug

```
<a class="pull-right" href="#&quot;" data-toggle="tooltip" data-html="true" data-bind="click: $root.toggleRelay, visible: $root.loginState.loggedIn(), tooltip: {title: label, placement: 'bottom'}" style="float: left;" data-original-title="" title="">
<i class="icon off icon-print" data-bind="css: [currentState(), icon(),(($root.processing().indexOf(ip()) > -1) ? 'icon-spin' : '')].join(' ')"></i>
**<span>plug name</span>**
</a>
```

and add this css :
```
#touch body #navbar .navbar-inner .container .nav li.dropdown#all_touchui_settings #navbar_plugin_tplinksmartplug a:before {
	display: none;
}

#touch body #navbar_plugin_tplinksmartplug > a > i {
	padding-left: 4px;
}

body #navbar_plugin_tplinksmartplug > a > span {
	display:none;
}

#touch body #navbar_plugin_tplinksmartplug > a > span {
	padding-left: 17px;
	display:inline;
}
```

I did not really test, only by hacking the HTML code.
but I can test, if you decide to implement it

**Comment by jneilliii** [comment]
Nice @devildant, I'll give that a try.  In the meantime I update the release that reverted the previous changes and should still fix the issue I was trying to solve.  Give it a go and let me know.

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16.zip

**Comment by devildant** [comment]
it's better, but I still have a problem with the plug state when loading the page.
reproduce:
1) activate the plug
2) refresh the page

the state of the plug changes from green to red.
the ajax checkstatus request returns the current state to "on".
I think the web-based socket has the status "off"

**Comment by jneilliii** [comment]
Does the status update properly once a polling interval is hit?

**Comment by jneilliii** [comment]
So integrated the touchui changes, but I don't have a 1.3.10 octoprint and as you already know I can't get touchui to load in 1.3.11rc3.  Install from here...

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16c.zip

**Comment by devildant** [comment]
> Does the status update properly once a polling interval is hit?
with polling with interval after 1 min  it's ok

**Comment by devildant** [comment]
> So integrated the touchui changes, but I don't have a 1.3.10 octoprint and as you already know I can't get touchui to load in 1.3.11rc3. Install from here...
> 
> https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16c.zip

yes i know, i also use the #touch for the moment
the creator of the plugin is not very active, it's been a long time since he does not correct his plugin.
I will test your version right away

**Comment by devildant** [comment]
there seems to be a problem with the plug button (it is no longer visible)
```
packed_core.js?79f3153d:15654 Error instantiating tplinksmartplugViewModel : TypeError: ko.observableDictionary is not a function
    at new tplinksmartplugViewModel (https://192.168.1.47/static/webassets/packed_plugins.js?55295a52:11771:36)
    at _createViewModelInstance (https://192.168.1.47/static/webassets/packed_core.js?79f3153d:15569:20)
    at HTMLDocument.<anonymous> (https://192.168.1.47/static/webassets/packed_core.js?79f3153d:15652:41)
    at mightThrow (https://192.168.1.47/static/webassets/packed_libs.js?ea9fbc9a:3535:29)
    at process (https://192.168.1.47/static/webassets/packed_libs.js?ea9fbc9a:3603:12)
```

**Comment by devildant** [comment]
otherwise for the icons of the other sidebar / tab it's ok
![image](https://user-images.githubusercontent.com/8700106/57187025-5391a500-6ee9-11e9-9627-af13165c6a26.png)

**Comment by devildant** [comment]
mmm, it seems that the whole viewmodel is in error

**Comment by devildant** [comment]
ok, you accidentally delete the file: ko.observableDictionary.js
and I think I found the cause of the plug button status problem, it's this callback : 
```
		self.onEventSettingsUpdated = function(payload) {
			if (self.settings.settings !== undefined && self.settings.settings.plugins !== undefined) {
				self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
			} 
		}
```
I wanted to use it on my plugin to see, and it's a disaster, in this reminder, my self is partially fed.

**Comment by devildant** [comment]
I confirm that it is onEventSettingsUpdated which poses problem.
just delete it

**Comment by devildant** [comment]
it misses the name in the "span" of the plug (icon)
```
<a class="pull-right" href="#" data-toggle="tooltip" data-html="true" data-bind="click: $root.toggleRelay, visible: $root.loginState.loggedIn(), tooltip: {title: label, placement: 'bottom'}" style="float: left;" data-original-title="" title="">
<i class="icon on icon-print" data-bind="css: [currentState(), icon(),(($root.processing().indexOf(ip()) > -1) ? 'icon-spin' : '')].join(' ')"></i>

<span data-bind-"text:="" label"=""></span>

</a>
```

**Comment by devildant** [comment]
here is a small improvement css
```
#touch body #navbar_plugin_tplinksmartplug > a > i {
    padding-left: 1px;
    font-size: 1.25rem!important;
}
```

**Comment by devildant** [comment]
if you want to test by yourself, on the page of octoprint F12 and console
and enter this cmd for force open  touch panel:
`
$("#touchui_settings_dialog").modal('show')`

**Comment by devildant** [comment]
and i detect 
![image](https://user-images.githubusercontent.com/8700106/57187854-b6893900-6ef5-11e9-8dc5-2897732d1c78.png)
I also detected a small problem of responsiveness with the bottom button, but it's not much

**Comment by jneilliii** [comment]
Ok, I reverted the onEventSettingsUpdated function back to it's original state because that should work, and I'm not sure why it would be causing errors for you in the first place.  I can't reproduce the error.  I've updated the css a bit for touchui and fixed the span issue.  The graph still doesn't work great on small screens, but it is there.  If the graph ends up being too narrow, it throws a lot of errors from plotly side.  Not much I can do about that, and would actually consider hiding the tab altogether in touchui rather than trying to resolve the issue.

**Comment by devildant** [comment]
hello @jneilliii 

I think I have found a better solution.
replace the callback "onEventSettingsUpdated" with "onSettingsBeforeSave"

```
		self.onEventSettingsUpdated = function(payload) {
			if (self.complete == true && self.settings.settings !== undefined && self.settings.settings.plugins !== undefined) {
				self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
			} 
		}
```
become :
```
		self.onSettingsBeforeSave = function(payload) {
			self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
                        self.checkStatuses(); // call check status for get current stat
		}
```

this has several advantages : 
1) this callback, is not calling the loading of the page, the get status is not override
2) this callback is called at the time of the setting's know, so it behaves like onEventSettingsUpdated but without the disadvantages
3) with the call to the checkstatus, it does not seem to be necessary to refresh the page after adding a plug

note: self.settings.settings.plugins.tplinksmartplug.arrSmartplugs() seems to have a plug status always off, that's what was wrong with "onEventSettingsUpdated". except that with this callback it was impossible to call the check status, because this callback was issued several times

**Comment by devildant** [comment]
I am currently testing this solution and it seems very good (and error free)

**Comment by devildant** [comment]
and if you want hide chart on touchui view use this css :
```

#touch body #tab_plugin_tplinksmartplug_link {
display:none;
}

#touch body #tab_plugin_tplinksmartplug_link2 {
display:none;
}

#touch body #tab_plugin_tplinksmartplug {
display:none;
}
```

**Comment by jneilliii** [comment]
Can you do me a favor and test this function instead and let me know how it goes.
```
		self.onSettingsBeforeSave = function(payload) {
                        self.checkStatuses(); // call check status for get current state
			self.settings.settings.plugins.tplinksmartplug.arrSmartplugs(self.arrSmartplugs()); // save settings with current states updated.
		}
```

**Comment by devildant** [comment]
hello, 
juste switch position of self.checkStatuses(); ?

**Comment by devildant** [comment]
ha ok i see, I think that this poses a problem when updating settings (add plug).
but i will try

**Comment by jneilliii** [comment]
It might not work the way I think it should, because of asynchronous calls of checkStatuses.

**Comment by devildant** [comment]
it's good that I thought, there is a problem if we add a plug.
just onSettingsBeforeSave  instead of onEventSettingsUpdated  work well

```
		self.onSettingsBeforeSave = function(payload) {
			self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
                        self.checkStatuses(); // call check status for get current stat
		}
```
with this callback the addition of a plug, the status changes to red and then green if the plug is on.
but nothing shocking

and there is no problem loading the page

**Comment by devildant** [comment]
in addition this change does not impact the logic of your code, and it is much more stable

**Comment by devildant** [comment]
in any case on my side it works really well

**Comment by jneilliii** [comment]
I ended up integrating your method with one slight change.  To keep the status checks down I am comparing the length of the two arrays and only checking statuses if they are not the same.  This should pick up on the addition of new plugs and prevent unneeded calls to the server during save operations.

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16c.zip

Let me know if you foresee any drawback to that approach.

**Comment by devildant** [comment]
perfect.
i changed the css,
so that the size of the icon does not affect the positioning
```
#touch body #navbar_plugin_tplinksmartplug > a > span {
    display: inline;
    position: absolute;
    left: 33px;
    line-height: 34px;
}
```

**Comment by devildant** [comment]
![image](https://user-images.githubusercontent.com/8700106/57197952-6f3e8f00-6f6d-11e9-94de-89db5b16f9be.png)
![image](https://user-images.githubusercontent.com/8700106/57197965-95fcc580-6f6d-11e9-8f93-adbc5f05e451.png)

**Comment by jneilliii** [comment]
@devildant, what is the resolution of the device you are using with TouchUI?  Trying to get an idea of what the minimum width is for the chart to graph properly and will assign the display:none style using the following. I think that way if it will work it will show and if it won't it will hide. You may have better insight into this media query stuff though?
```
@media (max-width: 800px) {
    #touch body #tab_plugin_tplinksmartplug_link {
        display:none;
    }

    #touch body #tab_plugin_tplinksmartplug_link2 {
        display:none;
    }

    #touch body #tab_plugin_tplinksmartplug {
        display:none;
    }
}
```

**Comment by devildant** [comment]
it's ok with this @media, my mobile have width = 414 but it's not optimal.
![image](https://user-images.githubusercontent.com/8700106/57252590-c7988e00-704c-11e9-992e-2c86be51ea99.png)
I think that Plotly may not be suitable for mobile, it would have to find a chart plugin more simple.
but hiding it on mobile is not messy

**Comment by jneilliii** [comment]
Interesting, when I use Chrome and emulate a small device like that in developer console plotly doesn't seem to be able to load. The graph gets squished to very narrow and all kinds of errors are thrown in console.

**Comment by devildant** [comment]
yes I have a lot of trouble too, I did a lot of f5 to display something correct.
without changing the chart, I think the media is the right solution

**Comment by jneilliii** [comment]
Cool, thanks.  I just need to figure out what the correct size is then and target that to hide this tab. Thanks for confirming for me.

**Comment by devildant** [comment]
I have a small request for functionality :p
would it be possible to have a button in the sidebar to force the update of the status (consumption)?

**Comment by jneilliii** [comment]
Give it some testing for me...added in the refresh button, the @media css to hide for touchui and smaller screens.  Let's see how that works.

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16d.zip

**Comment by devildant** [comment]
hello thx.
1) use this css (i increase left, it's better) : 
```
#touch body #navbar_plugin_tplinksmartplug > a > span {
	display: inline;
    position: absolute;
    left: 52px;
    line-height: 34px;
}
```

2) used the statement in this order the, as it is the case now, the "body #navbar_plugin_tplinksmartplug > a > span" can pose problem
```
#touch body #navbar_plugin_tplinksmartplug > a > i {
	padding-left: 1px;
	font-size: 1.25rem!important;
}

body #navbar_plugin_tplinksmartplug > a > span {
	display:none;
}
```
3) remove the if inside onSettingsBeforeSave, I had not seen before but if we want to change the icon it is necessary to refresh if there is the "if")
```
		self.onSettingsBeforeSave = function() {
			console.log('arrSmartplugs size changed, checking statuses');
			self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
			self.checkStatuses();
		}
```

4) little problem with the refresh button, when the plug is off the button is still activable, and with touchui the button is not visible, and when it is visible the click action no longer works (touch ui must clear the event at the header)
solution put the button inside "tplinksmartplug_sidebar.jinja2" (end)
```
<button class="btn btn-light btn-lg btn-block" data-bind="click: function() { checkStatuses(); }" title="Refresh"><i class="fa fa-refresh"></i> Refresh</button>
```
![image](https://user-images.githubusercontent.com/8700106/57489702-22cdb900-72b7-11e9-9fd2-66716acf2229.png)

![image](https://user-images.githubusercontent.com/8700106/57489801-7a6c2480-72b7-11e9-90e7-2079ce27028a.png)

my modif :
[OctoPrint-TPLinkSmartplug-0.9.16d.zip](https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/files/3164184/OctoPrint-TPLinkSmartplug-0.9.16d.zip)

for the problem of disabling the button, I did not find a simple solution because the status is specific to each plug, it would need a function that returns true if a plug is activated, and false if no plug is activated, with that I can disable for you the button with css and a small js

**Comment by jneilliii** [comment]
Thanks.  I'll work on number 3, I don't want to check statuses on every save, I think that just adds too much overhead.  Maybe this would work?
```
		self.onSettingsBeforeSave = function() {
			self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
			if(self.arrSmartplugs().length !== self.settings.settings.plugins.tplinksmartplug.arrSmartplugs().length){
				console.log('arrSmartplugs size changed, checking statuses');
				self.checkStatuses();
			}
		}
```
As for the refresh button, it checks all plugs statuses so disabling it if a plug is off doesn't make sense to me.  You only have one plug, but I have 5 now, and not only checks the energy status, but the on/off state as well.  I have no issue moving the button into the viewmodel as you recommended, will make that change. That actually makes it easier and I don't have to inject the button into the header like I'm doing now in the beforebinding call.

**Comment by devildant** [comment]
> Thanks. I'll work on number 3, I don't want to check statuses on every save, I think that just adds too much overhead. Maybe this would work?
> 
> ```
> 		self.onSettingsBeforeSave = function() {
> 			self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
> 			if(self.arrSmartplugs().length !== self.settings.settings.plugins.tplinksmartplug.arrSmartplugs().length){
> 				console.log('arrSmartplugs size changed, checking statuses');
> 				self.checkStatuses();
> 			}
> 		}
> ```
by doing this you override the local variable by the new value, so the if is useless.
Do not put any yew.
especially since this event is triggered before the saves, it is not every 5 min that we change the conf

```
		self.onSettingsBeforeSave = function() {
			console.log('arrSmartplugs size changed, checking statuses');
			self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
			self.checkStatuses();
		}
```
>As for the refresh button, it checks all plugs statuses so disabling it if a plug is off doesn't make sense to me. You only have one plug, but I have 5 now, and not only checks the energy status, but the on/off state as well. I have no issue moving the button into the viewmodel as you recommended, will make that change. That actually makes it easier and I don't have to inject the button into the header like I'm doing now in the beforebinding call.

Yes, I understood correctly, the idea that I exposed is the following one:
- if at least one plug is activated (accessible refresh button)
if all outlets are disabled (refresh button disabled)

**Comment by jneilliii** [comment]
Yeah, you're right about the onSettingsBeforeSave.   I guess I'll have to do the status check every time.

**Comment by jneilliii** [comment]
May have a solution to the status check using something like this maybe, testing now.
```
		self.onSettingsBeforeSave = function() {
			if(JSON.stringify(self.arrSmartplugs()) !== JSON.stringify(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs())){
				console.log('arrSmartplugs changed, checking statuses');
				self.arrSmartplugs(self.settings.settings.plugins.tplinksmartplug.arrSmartplugs());
				self.checkStatuses();
			}
		}
```

**Comment by devildant** [comment]
hello, 
i try same problem, if i remove the if it's ok... :(
I developed the function I was talking about, to disable or enable the refresh button

inside js : 

```
		self.allPlugIsDisable =  ko.computed(function() {
			var enablePlug = null;
			enablePlug = ko.utils.arrayFirst(self.arrSmartplugs(), function(item) {
				return item.currentState() == "on";
			});
			if (enablePlug == null)
				return false;
			return true;
		})

```

inside sidebar.jinja2
```
<button class="btn btn-light btn-lg btn-block" data-bind="click: function() { checkStatuses(); }, enable:allPlugIsDisable(); " title="Refresh"><i class="fa fa-refresh"></i> Refresh</button>
```

or try directly : 
[OctoPrint-TPLinkSmartplug-0.9.16d.zip](https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/files/3167289/OctoPrint-TPLinkSmartplug-0.9.16d.zip)

normally it's work, but i have only one plug :(

**Comment by jneilliii** [comment]
I think I might be close now.  Give this one a try and let me know if the if statement is working now.  I added some console logging to get more information from the browser as the process is happening.  When you save it should spit out the two arrays being compared.  I ended up switching from JSON.stringify to ko.toJSON because the embedded observables weren't exporting properly before. I incorporated the button disabling but had to modify it a bit for people with non-energy and energy monitoring plugs.

https://github.com/jneilliii/OctoPrint-TPLinkSmartplug/archive/0.9.16e.zip
